Design data logic rework (#5577)
holochain/holochain: docs/design/state_model.md

# Data State Model

## Overview

The Holochain data storage and validation architecture provides:

1. [Action and entry](./data_model.md) data storage for agents' authored chains.

> 1. Ops as the unit of validation, with an aggregated validity status for records.

Reviewer: duplicate `1.`
```suggestion
2. Ops as the unit of validation, with an aggregated validity status for records.
```

> 1. Ops as the unit of validation, with an aggregated validity status for records.

Reviewer: Record isn't defined in this document AFAICS.

Author: https://github.com/holochain/holochain/commit/04393db97789ccf90882c388b7d28162cb88dcbc

2. Validation limbo table (`LimboOp`) to track pending ops, with shared `DhtAction`/`DhtEntry` tables.
3. Unified data querying for both obligated and cached data in the DHT database.

## Architecture

### Core Principles

1. **Ops are the unit of validation**: All validation happens at the op level.
2. **Records aggregate op validity**: A record's validity is derived from its derived ops.
3. **Validation limbo isolates pending ops**: Unvalidated ops stay in `LimboOp` table until validated, with actions and entries in shared `DhtAction`/`DhtEntry` tables marked by NULL `record_validity`.
4. **Distinct schemas per database type**: Authored and DHT databases have schemas tailored to their needs.
5. **Unified data storage**: DHT database serves both obligated and cached data. Cached data can be distinguished by arc coverage as required.
6. **Clear state transitions**: Data moves through well-defined states with no ambiguity

### Database Structure

#### 1. Authored Database

> **Purpose**: Store an agent's own authored chain data.

Reviewer 2: I'll just note my preference that authored actions and ops are also stored in the dht database. I don't feel strongly if private entries are also stored in dht database in a separate table or stored in a separate database. I think this will melt away a lot of query complexity.

> **Purpose**: Store an agent's own authored chain data.

Author: @jost-s and @ThetaSinner agree to merge the authored data into the main database

Regardless of what data the agent stores and validates on behalf of other DHT agents, their own authored chain is 
always fully stored and accessible.

```sql
-- Authored actions
CREATE TABLE Action (
    hash         BLOB PRIMARY KEY,
    author       BLOB NOT NULL,
    seq          INTEGER NOT NULL,
    prev_hash    BLOB,
    timestamp    INTEGER NOT NULL,
    action_type  TEXT NOT NULL,
    action_data  BLOB,         -- Serialized ActionData enum, containing action-type fields
   
    -- Reference fields for entry meta
    entry_hash   BLOB,         -- NULL for non-entry actions
);

-- Authored entries
CREATE TABLE Entry (
    hash BLOB PRIMARY KEY,
    blob BLOB NOT NULL,
);

> -- Capability grants lookup table.

Reviewer 3: The action is just `CreateEntry` with a private [`Entry::CapGrant`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/entry/enum.Entry.html#variant.CapGrant) or [`Entry::CapClaim`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/entry/enum.Entry.html#variant.CapClaim).

--
-- For simpler querying of cap grants from the agent chain.
CREATE TABLE CapGrant (
    action_hash BLOB PRIMARY KEY,
    cap_access  TEXT NOT NULL, -- 'unrestricted', 'transferable', 'assigned'
    tag         TEXT,

    FOREIGN KEY(action_hash) REFERENCES Action(hash)
);

-- Capability claims table.
--
-- For recording cap claims from other agents that are granted to a local agent.
CREATE TABLE CapClaim (
    action_hash BLOB PRIMARY KEY,
    tag         TEXT NOT NULL,
    grantor     BLOB NOT NULL,
    secret      BLOB NOT NULL,
);

-- Chain lock table.
--
-- For coordinating countersigning sessions. Prevents new actions from being committed to the chain
-- while a countersigning session is in progress.
CREATE TABLE ChainLock (
    author                BLOB PRIMARY KEY,  -- Agent who holds the lock
    subject               BLOB NOT NULL,      -- What is being locked (e.g., session hash)
    expires_at_timestamp  INTEGER NOT NULL    -- Unix timestamp when lock expires
);

-- Authored ops.
-- 
-- For publishing state only, does not contain a complete op.

> CREATE TABLE AuthoredOp (

Reviewer 2: Why do we need to store ops in a separate database? Now that we first integrate, then publish, wouldn't it make more sense to publish from the dht database?

Author: Getting merged, as above. So this becomes simpler.

> CREATE TABLE AuthoredOp (

Reviewer 2: ```suggestion
CREATE TABLE AuthoredChainOp (
```

    hash        BLOB PRIMARY KEY,
    action_hash BLOB NOT NULL,
    op_type     TEXT NOT NULL,
    basis_hash  BLOB NOT NULL,
    
    -- Publishing state
    last_publish_time INTEGER,
    receipts_complete BOOLEAN,
    

>     FOREIGN KEY(action_hash) REFERENCES Action(hash)

Reviewer 2: I had a thought that we could add auto-incrementing integer id fields for the op and action tables, and then use those for cheaper join queries.

Author: https://github.com/holochain/holochain/commit/04393db97789ccf90882c388b7d28162cb88dcbc

);

-- Validation receipts for authored ops.
--
-- For tracking that other agents have validated our authored ops
CREATE TABLE ValidationReceipt (
    hash          BLOB PRIMARY KEY,
    op_hash       BLOB NOT NULL,
    validators    BLOB NOT NULL,
    signature     BLOB NOT NULL,
    when_received INTEGER NOT NULL,

    FOREIGN KEY(op_hash) REFERENCES AuthoredOp(hash)
);

-- Authored warrants.
--
-- For warrants issued by this agent about invalid behavior by other agents

> CREATE TABLE Warrant (

Reviewer 2: ```suggestion
CREATE TABLE AuthoredWarrantOp (
```

    hash      BLOB PRIMARY KEY,
    author    BLOB NOT NULL,
    timestamp INTEGER NOT NULL,
    warrantee BLOB NOT NULL,
    proof     BLOB NOT NULL,  -- Serialized WarrantProof (InvalidChainOp or ChainFork)

    -- Publishing state
    last_publish_time INTEGER,
    receipts_complete BOOLEAN
);
```

#### 2. DHT Database
**Purpose**: Store validated DHT data, and ops pending validation

```sql
-- DHT actions.
CREATE TABLE DhtAction (
   hash          BLOB PRIMARY KEY,
   author        BLOB NOT NULL,
   seq           INTEGER,      -- NULL for actions not in agent activity
   prev_hash     BLOB,         -- NULL for actions not in agent activity
   timestamp     INTEGER NOT NULL,
   action_type   TEXT NOT NULL,
   action_data   BLOB NOT NULL, -- Serialized ActionData enum

   -- Reference fields for entry meta
   entry_hash    BLOB,         -- NULL for non-entry actions

   -- Record validity (aggregated from all ops for this record)
   -- A record is the combination of action + entry (if applicable)
   -- NULL for records in limbo, 'valid' or 'rejected' after integration

>    record_validity TEXT -- NULL, 'valid', 'rejected'

Reviewer: I think it should either be `accepted` and `rejected` or `valid` and `invalid`.
```suggestion
   record_validity TEXT -- NULL, 'valid', 'invalid'
```

Author: Fixed

);

-- DHT entries.
CREATE TABLE DhtEntry (
   hash BLOB PRIMARY KEY,
   blob BLOB NOT NULL
);

-- Limbo for DHT ops which are in the process of being validated.
CREATE TABLE LimboOp (
    hash        BLOB PRIMARY KEY,
    op_type     TEXT NOT NULL,
    action_hash BLOB NOT NULL,

    -- DHT location
    basis_hash         BLOB NOT NULL,
    storage_center_loc INTEGER NOT NULL,

    -- Local validation state
    validation_stage      TEXT NOT NULL, -- 'pending_sys', 'pending_app', 'complete'

>     sys_validation_status TEXT,          -- NULL, 'valid', 'rejected', 'abandoned'

Reviewer 2: Why text versus int?

Author: Good point https://github.com/holochain/holochain/commit/04393db97789ccf90882c388b7d28162cb88dcbc

    app_validation_status TEXT,          -- NULL, 'valid', 'rejected', 'abandoned'

    -- Validation receipt requirement
    require_receipt BOOLEAN NOT NULL,    -- Whether to send validation receipt back to author

    -- Timing and attempt tracking
    when_received INTEGER NOT NULL,
    sys_validation_attempts INTEGER DEFAULT 0,
    app_validation_attempts INTEGER DEFAULT 0,
    last_validation_attempt INTEGER,

    -- Storage tracking
    serialized_size INTEGER NOT NULL,  -- Size in bytes, calculated when op arrives

    FOREIGN KEY(action_hash) REFERENCES DhtAction(hash)
);

-- DHT ops which have completed validation and are integrated into the DHT.
CREATE TABLE DhtOp (
    hash        BLOB PRIMARY KEY,

>     op_type     TEXT NOT NULL,

Reviewer 2: What are the options for this? Does it correspond to ChainOpType or DhtOpType or both? Could the name match the rust type?

    -- DHT location
    basis_hash         BLOB NOT NULL,
    storage_center_loc INTEGER NOT NULL,

    -- Final validation result
    validation_status TEXT NOT NULL,    -- 'valid', 'rejected'
    locally_validated BOOLEAN NOT NULL, -- whether this op validated by us, or fetched from an authority

    -- Timing
    when_received   INTEGER NOT NULL, -- copied from LimboOp
    when_integrated INTEGER NOT NULL, -- set when moved from LimboOp

    -- Storage tracking
    serialized_size INTEGER NOT NULL DEFAULT 0, -- size in bytes for storage quota management

    FOREIGN KEY(action_hash) REFERENCES DhtAction(hash)
);

-- DHT warrants.
--
-- For warrants received from the network about invalid behavior
CREATE TABLE DhtWarrant (
    hash      BLOB PRIMARY KEY,
    author    BLOB NOT NULL,
    timestamp INTEGER NOT NULL,
    warrantee BLOB NOT NULL,
    proof     BLOB NOT NULL,  -- Serialized WarrantProof (InvalidChainOp or ChainFork)

    -- DHT location (stored at warrantee's agent authority)
    storage_center_loc INTEGER NOT NULL
);

-- Link index table.
--
-- For efficient link queries. Populated at integration time when CreateLink ops are validated.

> CREATE TABLE Link (

Author: The only alternative as the other tables are defined would be queries by action type, followed by filtering in memory, which definitely won't perform well. I find this clearer than duplicated fields in a single table with many indexes but I'm open to suggestions for this.

CREATE TABLE UpdatedRecord (
    action_hash              BLOB PRIMARY KEY,  -- The Update action
    original_action_hash     BLOB NOT NULL,      -- The original Create or Update being updated
    original_entry_hash      BLOB NOT NULL,      -- The original entry being updated

    FOREIGN KEY(action_hash) REFERENCES DhtAction(hash) ON DELETE CASCADE
);

CREATE INDEX idx_updated_record_original ON UpdatedRecord(original_action_hash);

-- Delete index table.
--
-- For efficient delete queries. Populated at integration time when Delete actions are validated.
CREATE TABLE DeletedRecord (
    action_hash         BLOB PRIMARY KEY,  -- The Delete action
    deletes_action_hash BLOB NOT NULL,      -- The action being deleted
    deletes_entry_hash  BLOB NOT NULL,      -- The entry being deleted

    FOREIGN KEY(action_hash) REFERENCES DhtAction(hash) ON DELETE CASCADE
);

CREATE INDEX idx_deleted_record_deletes ON DeletedRecord(deletes_action_hash);
```

### Rust Structure

Actions:

```rust
/// Common action header stored for all action types
pub struct ActionHeader {
    pub author: AgentPubKey,
    pub timestamp: Timestamp,
    pub action_seq: u32,

>     pub prev_action: ActionHash,

Author: https://github.com/holochain/holochain/commit/04393db97789ccf90882c388b7d28162cb88dcbc

}

/// Action-specific data stored separately from header
pub enum ActionData {
    Dna(DnaData),

>     AgentValidationPkg(AgentValidationPkgData),

Reviewer 2: Could potentially merge these two

Reviewer 2: I think I would lean towards merging them, since they are always committed together.

Author: As part of the discussion about DNA migration, we talked about using these independently. I'm just going to leave this for now.

    InitZomesComplete(InitZomesCompleteData),
    Create(CreateData),
    Update(UpdateData),
    Delete(DeleteData),
    CreateLink(CreateLinkData),
    DeleteLink(DeleteLinkData),
}

> /// Full action with data loaded on-demand

Reviewer: loaded on demand? `data` is not an optional field, so what does on deman mean here?

pub struct DeleteData {
    pub deletes_address: ActionHash,
    pub deletes_entry_address: EntryHash,
    pub weight: RateWeight,
}

pub struct CreateLinkData {
    pub base_address: AnyLinkableHash,
    pub target_address: AnyLinkableHash,
    pub zome_index: ZomeIndex,
    pub link_type: LinkType,
    pub tag: LinkTag,
    pub weight: RateWeight,
}

pub struct DeleteLinkData {
    pub base_address: AnyLinkableHash,
    pub link_add_address: ActionHash,
}

// Minimal data for chain-only actions
pub struct DnaData {
    pub dna_hash: DnaHash,
}

pub struct AgentValidationPkgData {
    pub membrane_proof: Option<MembraneProof>,
}

pub struct InitZomesCompleteData {}
```

DHT Operations:

```rust
/// Top-level DHT operation that can be either a chain operation or a warrant.
pub enum DhtOp {
    /// An op representing storage of some record information.
    ChainOp(Box<ChainOp>),
    /// An op representing storage of a claim that a ChainOp was invalid.
    WarrantOp(Box<WarrantOp>),
}

/// Represents how entry data is included in an op.
pub enum OpEntry {
    /// The entry is present in this op
    Present(Entry),
    /// The action references a private entry, which is not included

>     Hidden,

Reviewer 2: ```suggestion
    Public(Entry),
    /// The action references a private entry which I have authored
    AuthoredPrivate(Entry),
    /// The action references a private entry which I have not authored
    Private
```

I'd prefer consistent meaning no matter the context the struct is used

}

/// Chain operations that represent chain data distributed across the network.
/// Each operation is stored at a specific DHT location determined by its basis hash.

> pub enum ChainOp {

Reviewer 2: In a sharded network there are 4 distinct types of locations:
- A location that stores a Record
- A location that stores all actions by an Author
- A location that stores all state changes for a specific Entry
- A location that stores all state changes to a specific Link

All of the ChainOp variants outlined here store one of the following data structures:
- Action + Entry
- Action

When we receive a ChainOp, we want to run all sys validation logic that we are able to on that data. Some sys validation logic depends on other data, so we want to run all the logic that depends on data we expect to be storing already, or are willing to fetch because we want to store it.

To do this, we can look at the ChainOp, determine its locations of each of the 4 types, and check if they fall within our arc. Then we can run the validation logic corresponding to those location types.

In the current approach, we need to actually fetch a different ChainOp in order to run the validation logic for that type of location, even when it contains data that we already have. Additionally we open ourselves up to an unnecessary vulnerability where certain ChainOp types are never published and therefore their validation logic is never run.

I hope this makes some sense, it is what I'm trying to articulate with https://github.com/holochain/holochain/issues/5487

Reviewer 2: I think the linked issue describes how it fits into k2 though it could probably be clearer. K2 gossip tracks a separate redundant op id for each [hash + location u32 where it is stored], but then fetch queue still just fetches by the hash. When holochain receives the op, it retrieves the validation type locations from the op itself, and then runs validations on all the types where the location falls within its arc.

Author: Added a section about expanding records to other op types locally which may save work. I think that was meant to be implemented already but I'm pretty sure it's not properly implemented in the workflows.

Beyond that, I'm still not convinced we can do this without changes to K2.

    /// 
    /// Only created if the original entry was public.
    DeleteEntry(SignedAction),
    /// Deletes indexed at the original record authority.
    DeleteRecord(SignedAction),
    /// Links indexed at the base address.
    CreateLink(SignedAction),
    /// Link deletes indexed at the base address.
    DeleteLink(SignedAction),
}

/// Warrant operation representing a claim that a ChainOp was invalid.
pub struct WarrantOp(SignedWarrant);

/// Internal representation of a ChainOp after hash and signature verification.
///
/// This type is used internally after counterfeit checks have been completed. 
/// The hashes are verified once during the incoming ops workflow and retained.
pub(crate) struct HashedChainOp {
    /// The verified op hash
    pub op_hash: DhtOpHash,
    /// The signed action with verified hash
    pub action: SignedActionHashed,
    /// The entry with verified hash (if present in the op)
    pub entry: Option<EntryHashed>,
    /// The type of chain operation
    pub op_type: ChainOpType,
    /// The DHT location where this op is stored
    pub basis_hash: AnyDhtHash,
    /// The numeric storage center location (derived from basis_hash)

>     pub storage_center_loc: u32,

Reviewer: Why are there both these location fields? For convenience?

}
```

### Creation and Distribution Flow

The high-level flow for authoring actions and distributing ops is as follows:

```
1. Self validation
   ├─> Run sys validation check
   ├─> Run app validation checks via WASM
   └─> On failure: rollback transaction, chain unchanged

2. Author new action locally
   ├─> Insert into `Action` table
   ├─> Insert into `Entry` table (if applicable)
   ├─> Insert into `CapGrant` or `CapClaim` table (if applicable)
   ├─> Insert into `Link` table (if action is CreateLink)

>    └─> Insert into `DeletedLink` table (if action is DeleteLink)

Author: I was wrong, not here. This is the authored database which doesn't have those index tables.

     - Timestamp breaks ties when multiple ops reference the same action
     - Simpler than maintaining a computed ordering field
   - **Resolution**: No `op_order` column needed in `AuthoredOp`

### Validation Flow

The validation flow processes incoming DHT ops through several stages, from initial receipt through final integration into the DHT database.

```
1. Incoming DHT Ops Workflow
   ├─> Compute op hash (`DhtOpHash`) from op content
   ├─> Filter duplicate ops already being processed
   ├─> Verify counterfeit checks (signature and hash verification)
   ├─> Convert `ChainOp` to `HashedChainOp` (internal type with checked hashes)
   ├─> Filter ops already in database (check `DhtOp` table)
   ├─> Insert action into `DhtAction` with `record_validity=NULL`
   ├─> Insert entry (if applicable) into `DhtEntry`
   └─> Insert into `LimboOp` with `validation_stage='pending_sys'`

2. Sys Validation Workflow
   ├─> Query `LimboOp` for ops with `validation_stage='pending_sys'`
   ├─> Check dependencies in `DhtAction`/`DhtEntry`
   ├─> Perform sys validation checks
   └─> Update `sys_validation_status` in `LimboOp`

>        └─> Trigger app validation if valid

Reviewer: When invalid, author warrant and trigger integration?

    - Query `DhtOp` table to filter ops already in database
   ```sql
   SELECT EXISTS(
       SELECT 1 FROM DhtOp WHERE DhtOp.hash = :hash
   )
   ```
    - Skip ops that already exist

4. **Deduplication Check**
   - Check the current working state to filter ops already being processed to prevent duplicate work for ops in flight.

5. **Counterfeit Checks and Hash Verification**
   - For each op, verify structural integrity and cryptographic validity:
     - **Signature verification**:
         - For `ChainOp`: `verify_action_signature(signature, action)` ensures signature is valid for the action
         - For `WarrantOp`: `verify_warrant_signature(warrant_op)` ensures warrant signature is valid
     - **Action hash verification**: Using `holo_hash` to compute the action hash ensures that the action content matches the provided action hash.
     - **Entry hash verification**: For actions with an entry, use `holo_hash` to verify entry content hashes to the entry hash referenced in the action.
   - **Convert to HashedChainOp**: After verification, convert each `ChainOp` to `HashedChainOp` containing the verified hashes. This avoids re-computing hashes during database insertion.

>    - **Batch rejection**: If any op fails any check, the entire incoming batch is dropped

Reviewer 2: So this is fine, but just to flag it I think it could be a griefing vector that a peer can send us malformed ops, and we have no way to warrant or block that peer.

     - This prevents the invalid op from entering the system

>      - It also prevents wasting resources processing other ops in the batch

Reviewer 2: Why would we assume that other ops in batch would be bad as well?

    AND (LimboOp.app_validation_status = 'valid' OR LimboOp.app_validation_status = 'rejected')
ORDER BY LimboOp.when_received
```

**Workflow Steps:**

1. **Query Completed Ops**
   - Select ops with `validation_stage = 'complete'` and both sys and app validation statuses set.
   - Both valid and rejected ops are integrated.

2. **Move Op to DhtOp Table**
   - Start a write transaction for each op.
   - Execute the following insert:
   ```sql
   INSERT INTO DhtOp (
       hash,
       op_type,
       action_hash,
       basis_hash,
       storage_center_loc,
       validation_status,
       locally_validated,
       when_received,
       when_integrated,
       serialized_size
   )
   SELECT
       hash,
       op_type,
       action_hash,
       basis_hash,
       storage_center_loc,
       CASE
           WHEN sys_validation_status = 'valid' AND app_validation_status = 'valid' THEN 'valid'
           ELSE 'rejected'
       END,
       TRUE,  -- locally_validated
       when_received,
       CURRENT_TIMESTAMP,
       serialized_size  -- calculated when op arrived
   FROM LimboOp
   WHERE hash = :op_hash
   ```

3. **Update `DhtAction` with Record Validity**

>    - Aggregate `record_validity` from all ops for this action using a single query

Author: These are expected consequences of this change. The way we're doing things now is more problematic. For example, a `must_get_valid_record` that retrieves a record that is valid for its `StoreRecord` op but invalid for its `RegisterAgentActivity` is just ignoring the actual validity of the record. With sharding, we do just have partial knowledge and the best we can do is use the information we have locally until we learn more.

This is opening a can of worms in a sense, because there will be new functionality needed to deal with the consequences of this change but what we have right now is not what was designed for the system originally and is really counterintuitive for app developers - to a point where it's really not practical to assess whether the app validation logic inside an app is coherent.

   ```sql
   DELETE FROM DeletedRecord
   WHERE action_hash = :action_hash
   AND :action_hash IN (
       SELECT hash FROM DhtAction
       WHERE hash = :action_hash AND record_validity = 'rejected'
   );
   ```

5. **Insert Into Warrant Table** (if applicable)
   - If the op is a warrant op (`op_type = 'Warrant'`) and the op is valid, insert into `DhtWarrant` table:
   ```sql
   INSERT INTO DhtWarrant (hash, author, timestamp, warrantee, proof, storage_center_loc)
   VALUES (
       :warrant_hash,
       :author,        -- extracted from warrant proof
       :timestamp,     -- extracted from warrant
       :warrantee,     -- extracted from warrant
       :proof,         -- serialized WarrantProof BLOB
       :storage_center_loc
   )
   ```

6. **Delete from `LimboOp`**
   ```sql
   DELETE FROM LimboOp WHERE hash = :op_hash
   ```

7. **Commit Transaction**
    - Commit the write transaction to finalize changes for this op.

8. **Send Validation Receipt** (if required)
   - If op came from network and `require_receipt = true`
   - Send a signed validation receipt back to the author

> ### Record Validity Aggregation

Author: Chain fork warrants only affect the status of the author. As long as both forks are independently valid - we don't modify our validity state. We can't know which fork we'll choose at that point, so we can't make record validity decisions.

**Rules:**
1. A record is **INVALID** if ANY known ops for it are rejected.
2. A record is **VALID** if at least one known op for it is valid and no known ops are rejected.
3. Records in limbo have a `NULL` value for `record_validity`. After integration of the first op for a record, `record_validity` is `'valid'` or `'rejected'`.

> The record validity is determined at the time of integration by examining all known ops associated with the record. Due to the sharding model, a validator may not have all ops for a record - they hold specific op types over specific ranges. This aggregated status is stored with the record itself, eliminating the need for complex joins during queries.

Reviewer: It reads like all ops of a record are queried when a record is integrated. I believe that with every op integration, the record validity is updated if the op is invalid.
```suggestion
The record validity is first set at the time of integrating the first of its ops. If subsequent op validations find an op to be invalid, the record validity is updated to ``rejected``. Due to the sharding model, a validator may not have all ops for a record - they hold specific op types over specific ranges. This aggregated status is stored with the record itself, eliminating the need for complex joins during queries.
```

> ### Record Validity Correction

Author: See above, and we can discuss

TODO: It is a future piece of work to define this logic. When implemented, this logic must also clean up index tables (`Link`, `DeletedLink`, `UpdatedRecord`, `DeletedRecord`, `CapGrant`, `CapClaim`) when records are invalidated.

## Query Patterns

> ### Get Record by Hash

Reviewer 2: We should add a section for must_get_valid_record. I don't think it can be a standard get query anymore or be based on record_validity, it needs to specify the specific op type(s) that must be valid to be deterministic.

Reviewer 2: I think this is the same concept of record validity as as above so we can continue it there.

Just to note again that we got a bit further along in the discussion in this MM thread, which would be good to revisit: https://chat.holochain.org/holo/pl/yyxuwgycwj84iq8r18y45f7nwe

Retrieve a complete record (action and entry) from the DHT by action hash.

```sql
SELECT Action.*, Entry.*
FROM DhtAction AS Action
LEFT JOIN DhtEntry AS Entry ON Action.entry_hash = Entry.hash
WHERE Action.hash = ?
  AND Action.record_validity = 'valid'
```

### Get Entry by Hash

Retrieve an entry from the DHT using the validation status from any action that references it.

```sql
SELECT Entry.*, Action.*
FROM DhtEntry AS Entry
JOIN DhtAction AS Action ON Action.entry_hash = Entry.hash
WHERE Entry.hash = ?
  AND Action.record_validity = 'valid'
LIMIT 1
```

Justification for *any*: An entry can be referenced by multiple actions. If at least one action referencing the entry is valid, the entry is considered valid to be served to the network or used.

### Get Links

Query all non-deleted links from a base hash using the `Link` index table populated at integration time.

Find non-deleted links from base:

```sql
SELECT Link.*, Action.*
FROM Link
JOIN DhtAction AS Action ON Link.action_hash = Action.hash
LEFT JOIN DeletedLink ON Link.action_hash = DeletedLink.create_link_hash
WHERE Link.base_hash = ?
  AND DeletedLink.create_link_hash IS NULL

>   AND Action.record_validity = 'valid'

Reviewer 2: Any default ordering?

Count non-deleted links:

```sql
SELECT COUNT(*)
FROM Link
LEFT JOIN DeletedLink ON Link.action_hash = DeletedLink.create_link_hash
JOIN DhtAction AS Action ON Link.action_hash = Action.hash
WHERE Link.base_hash = ?
  AND DeletedLink.create_link_hash IS NULL
  AND Action.record_validity = 'valid'
```

Additional filters (link_type, tag prefix, author, timestamp) applied as needed.

### Get Agent Activity

Query an agent's chain activity with flexible filtering for action types, entry types, and validation status.

```sql
SELECT Action.*
FROM DhtAction AS Action
WHERE Action.author = ?
  AND Action.record_validity IS NOT NULL
ORDER BY Action.seq
```

Application code filters for:
- Specific action_type values
- Entry types (via action_data deserialization)
- Sequence ranges
- Include valid/rejected/warrants based on GetActivityOptions

### Must Get Agent Activity

Deterministic hash-bounded query for agent activity used in countersigning scenarios.

Query with sequence bounds:

```sql
SELECT Action.*
FROM DhtAction AS Action
WHERE Action.author = ?
  AND Action.seq BETWEEN ? AND ?
  AND Action.record_validity IS NOT NULL
ORDER BY Action.seq
```

> Or bounded by prev_hash chain traversal.

Author: https://github.com/holochain/holochain/commit/04393db97789ccf90882c388b7d28162cb88dcbc

FROM Action
LEFT JOIN Entry ON Action.entry_hash = Entry.hash
WHERE Action.author = ?
  AND Action.seq BETWEEN :start_seq AND :end_seq
ORDER BY Action.seq ASC
```

**ActionHashRange(start_hash, end_hash)**: Hash-bounded range with fork disambiguation
- Query all actions
- Walk backwards from `end_hash` following `prev_hash` until reaching `start_hash`
- Application code performs the chain traversal (cannot be done efficiently in SQL)

**ActionHashTerminated(end_hash, n)**: Hash-terminated with N preceding records
- Query all actions
- Walk backwards from `end_hash` following `prev_hash` for N steps
- Application code performs the chain traversal

**Entry Type Filtering:**

Entry type filtering requires deserializing `action_data` to extract the `entry_type` field. This should be handled in application code after fetching matching actions,

### Cap Grant and Claim Lookups

Cap grant/claim lookups use dedicated tables for direct access without chain scans:

Find cap grant by access type (direct lookup, join for other fields):

```sql
SELECT cg.*, Action.*, Entry.*
FROM CapGrant cg
JOIN Action ON cg.action_hash = Action.hash
JOIN Entry ON Action.entry_hash = Entry.hash

> WHERE cg.cap_access = ?;

Author: https://github.com/holochain/holochain/pull/5577/commits/04393db97789ccf90882c388b7d28162cb88dcbc

```sql
SELECT * FROM Action WHERE hash = ?;
```

Then deserialize `action_data` BLOB in application.

### Find Updates for a Create Action

Retrieve all `Update` actions that reference a specific `Create` action. Updates form a chain where each `Update` references either the original `Create` or another `Update`.

Uses the `UpdatedRecord` index table populated during integration.

```sql
-- Find direct updates (one hop)
SELECT Action.*, Entry.*
FROM UpdatedRecord
JOIN DhtAction AS Action ON UpdatedRecord.action_hash = Action.hash
LEFT JOIN DhtEntry AS Entry ON Action.entry_hash = Entry.hash
WHERE UpdatedRecord.original_action_hash = ?
  AND Action.record_validity = 'valid'
ORDER BY Action.seq

-- Find full update chain (recursive)
WITH RECURSIVE update_chain AS (
  -- Base case: direct updates of the create
  SELECT
    UpdatedRecord.action_hash as hash,
    Action.author,
    Action.seq,
    Action.timestamp,
    Action.entry_hash,
    1 as depth
  FROM UpdatedRecord
  JOIN DhtAction AS Action ON UpdatedRecord.action_hash = Action.hash
  WHERE UpdatedRecord.original_action_hash = ?
    AND Action.record_validity = 'valid'

  UNION ALL

>   -- Recursive case: updates of updates

Reviewer 2: We've discussed dropping update chains and having all updates be on the create action. I assume we don't want to include that in this stability work?

Author: I'm slightly undecided about that. After more time thinking it through, I do see a benefit to allowing updates to updates. It just makes our code so much more complicated. I have chosen not to change that for now but I'm open to discussion on it

4. **Single database per DNA**: Each DNA cell uses one database for all chain, DHT, and validation data. Private entries are isolated in a dedicated table for access control auditing.
5. **Unified data storage**: The database serves authored chain data, obligated DHT data, and cached data. Cached data can be distinguished by arc coverage as required.
6. **Clear state transitions**: Data moves through well-defined states with no ambiguity.
7. **Action-type specific handling**: Different action types (Create, Update, Delete, Link) have specific handling rules for constructing a DHT state that is queryable.

### Database Structure

#### Per-DNA Database

Each DNA cell uses a single database for authored chain data, DHT data, and validation state.

```sql
-- Actions (stores both locally authored and network-received actions).
--
-- WITHOUT ROWID stores rows in a single B-tree clustered on hash, eliminating the
-- double-lookup overhead of a regular table (rowid B-tree + primary key index).
-- All FK references already store the full 32-byte hash, so no downstream changes needed.

> CREATE TABLE DhtAction (

Author: I think we could remove the prefixes to these table names. They were in place to ensure that DHT and authored database queries were kept separate but we no longer need a `Dht` prefix on table names.

```

**Differences to Current Implementation (Database Structure):**

1. **Merged authored and DHT databases**: The current implementation uses separate per-agent authored and per-DNA DHT databases. The new design uses a single per-DNA database. For full-arc nodes (the common case), every authored write already results in a DHT write, so the separate write handle provides no meaningful performance benefit. The single database eliminates cross-database query complexity.

2. **Private entries in dedicated table**: Private entries move from the general `Entry` table to a dedicated `PrivateEntry` table with an `author` column. This makes access control auditable at the schema level — code that queries `DhtEntry` cannot accidentally leak private content.

3. **Self-authored ops bypass limbo**: Self-authored ops are pre-validated at authoring time and inserted directly into `DhtChainOp` with `validation_status = 1`. The existing duplicate check in the incoming ops workflow (`SELECT EXISTS(SELECT 1 FROM DhtChainOp WHERE hash = :hash)`) prevents double-processing when the same op arrives from the network.

> 4. **Publishing state on integrated tables**: Publishing fields (`last_publish_time`, `receipts_complete`) move from the removed `AuthoredChainOp` table onto `DhtChainOp`. `DhtWarrant` gets `last_publish_time` only (no receipt tracking for warrants). These fields are NULL for network-received ops and only populated for locally authored ops.

Author: What does this mean for table size? Should we move these author-only fields into a separate table to avoid having a lot of zeroes in them for non-local agents?

### Record Validity Aggregation

**Rules:**
1. A record is **rejected** (`record_validity = 2`) if ANY known ops for it are rejected.
2. A record is **accepted** (`record_validity = 1`) if at least one known op for it is accepted and no known ops are rejected.
3. Self-authored records are inserted with `record_validity = 1` (pre-validated at authoring time).
4. Network-received records in limbo have a `NULL` value for `record_validity`. After integration of the first op for a record, `record_validity` is `1` (accepted) or `2` (rejected).

> The record validity is first set at authoring time for self-authored records, or at the time of integrating the first of its ops for network-received records. If later op integrations find an op to be rejected, the record validity is updated to `2` (rejected). Due to the sharding model, a validator may not have all ops for a record — they hold specific op types over specific ranges. This aggregated status is stored with the record itself, eliminating the need for complex joins during queries.

Author: We still need to run integration for self-authored ops currently because part of integration is telling the DHT model about ops that we want included in the model.
